### PR TITLE
[Promotion] Discount based on items from specific taxons

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT
| Doc PR          |

- [x] scenarios for taxon rule (with multiple taxons choice)
- [x] new steps implementation

Fix from separate PR: https://github.com/Sylius/Sylius/pull/4595
- [x] valid taxon rule

### CODE REVIEW

File: src/Sylius/Behat/Context/Transform/TaxonContext.php [hunk 1]
```diff
@@ -45,4 +45,15 @@ public function getTaxonByName($taxonName)
 
         return $taxon;
     }
+
+    /**
+     * @Transform classified as :firstTaxon or :secondTaxon
```
Comment (pamil) [inline src/Sylius/Behat/Context/Transform/TaxonContext.php]: I'm not sure about this transformer.

File: src/Sylius/Component/Core/Promotion/Checker/TaxonRuleChecker.php [hunk 2]
```diff
@@ -36,13 +42,13 @@ public function isEligible(PromotionSubjectInterface $subject, array $configurat
         /* @var $item OrderItemInterface */
         foreach ($subject->getItems() as $item) {
             foreach ($item->getProduct()->getTaxons() as $taxon) {
-                if ($configuration['taxons']->contains($taxon->getId())) {
-                    return !$configuration['exclude'];
+                if (in_array($taxon->getId(), $configuration['taxons'])) {
```
Comment (pamil) [inline src/Sylius/Component/Core/Promotion/Checker/TaxonRuleChecker.php]: Use `true` as the third parameter to do a strict comparison.

File: src/Sylius/Component/Core/spec/Promotion/Checker/TaxonRuleCheckerSpec.php [hunk 3]
```diff
@@ -53,16 +55,13 @@ function it_should_recognize_subject_as_eligible_if_product_taxon_is_matched(
...
     }
 
-    function it_should_recognize_subject_as_not_eligible_if_product_taxon_is_not_matched(
+    function it_recognizes_subject_as_not_eligible_if_product_taxon_is_not_matched(
         OrderInterface $subject,
         OrderItemInterface $item,
-        Taxon $taxon,
         Product $product,
-        ArrayCollection $collection
+        Taxon $taxon
```
Comment (pamil) [inline src/Sylius/Component/Core/spec/Promotion/Checker/TaxonRuleCheckerSpec.php]: TaxonInterface?

File: src/Sylius/Component/Core/spec/Promotion/Checker/TaxonRuleCheckerSpec.php [hunk 4]
```diff
@@ -53,16 +55,13 @@ function it_should_recognize_subject_as_eligible_if_product_taxon_is_matched(
         $this->isEligible($subject, $configuration)->shouldReturn(true);
     }
 
-    function it_should_recognize_subject_as_not_eligible_if_product_taxon_is_not_matched(
+    function it_recognizes_subject_as_not_eligible_if_product_taxon_is_not_matched(
         OrderInterface $subject,
         OrderItemInterface $item,
-        Taxon $taxon,
         Product $product,
```
Comment (pamil) [inline src/Sylius/Component/Core/spec/Promotion/Checker/TaxonRuleCheckerSpec.php]: ProductInterface?

File: src/Sylius/Component/Core/spec/Promotion/Checker/TaxonRuleCheckerSpec.php [hunk 5]
```diff
@@ -29,21 +34,18 @@ function it_is_initializable()
...
     {
         $this->shouldImplement(RuleCheckerInterface::class);
     }
 
-    function it_should_recognize_subject_as_eligible_if_product_taxon_is_matched(
+    function it_recognizes_subject_as_eligible_if_product_taxon_is_matched(
         OrderInterface $subject,
         OrderItemInterface $item,
-        Taxon $taxon,
         Product $product,
```
Comment (pamil) [inline src/Sylius/Component/Core/spec/Promotion/Checker/TaxonRuleCheckerSpec.php]: ProductInterface?

File: src/Sylius/Behat/Context/Transform/TaxonContext.php (hunk 1 above)
Comment (Zales0123) [inline src/Sylius/Behat/Context/Transform/TaxonContext.php]: I'm not sure how to do it better ;/

File: src/Sylius/Component/Core/spec/Promotion/Checker/TaxonRuleCheckerSpec.php [hunk 6]
```diff
@@ -29,21 +34,18 @@ function it_is_initializable()
...
-        ArrayCollection $collection
+        Taxon $taxon
```
Comment (pamil) [inline src/Sylius/Component/Core/spec/Promotion/Checker/TaxonRuleCheckerSpec.php]: TaxonInterface?

File: features/promotion/receiving_discount_based_on_items_from_specific_taxons.feature [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        Given the promotion gives "€50.00" off if order contains products classified as "T-Shirts"
+        When I add 3 products "PHP T-Shirt" to the cart
+        And my cart total should be "€250.00"
+        And my discount should be "-€50.00"
+
+    @ui
+    Scenario: Receiving discount on order while buying product from both of promoted taxon
+        Given the promotion gives "€10.00" off if order contains products classified as "T-Shirts"
+        And there is a promotion "Mugs promotion"
+        Given the promotion gives "€5.00" off if order contains products classified as "Mugs"
```
Comment (pjedrzejewski) [inline features/promotion/receiving_discount_based_on_items_from_specific_taxons.feature]: And this promotion? We have duplicated Given here.

**Comment by pjedrzejewski** [comment]
Thanks Mateusz!

**Pull request merged**